win32ss/gdi/ntgdi/dibobj.c

```diff
@@ -1242,147 +1250,243 @@ NtGdiStretchDIBitsInternal(
+        rcSrc.right = xSrc + abs(cxSrc);
+        rcSrc.bottom = ySrc + abs(cySrc);
+        rcDst.left = xDst;
+        rcDst.top = yDst;
+        rcDst.right = rcDst.left + cxDst;
+        rcDst.bottom = rcDst.top + cyDst;
+        IntLPtoDP(pdc, (POINTL*)&rcDst, 2);
+        RECTL_vOffsetRect(&rcDst, pdc->ptlDCOrig.x, pdc->ptlDCOrig.y);
+
+        if (pdc->fs & (DC_ACCUM_APP|DC_ACCUM_WMGR))
+        {
+           IntUpdateBoundsRect(pdc, &rcDst);
+        }
+
+        hbmTmp = GreCreateBitmapEx(pbmi->bmiHeader.biWidth,
+                                   abs(pbmi->bmiHeader.biHeight),
+                                   0,
+                                   BitmapFormat(pbmi->bmiHeader.biBitCount,
+                                                pbmi->bmiHeader.biCompression),
+                                   pbmi->bmiHeader.biHeight < 0 ? BMF_TOPDOWN : 0,
+                                   cjMaxBits,
+                                   safeBits,
+                                   0);
+
+        if (!hbmTmp)
+        {
+            goto cleanup;
+        }
+
+        psurfTmp = SURFACE_ShareLockSurface(hbmTmp);
+        if (!psurfTmp)
+        {
+
+            goto cleanup;
+        }
+
+        /* Create a palette for the DIB */
+        ppalDIB = CreateDIBPalette(pbmi, pdc, dwUsage);
+        if (!ppalDIB)
+        {
+            goto cleanup;
+        }
+
+        /* Prepare DC for blit */
+        DC_vPrepareDCsForBlit(pdc, &rcDst, NULL, NULL);
+
+        psurfDst = pdc->dclevel.pSurface;
+
+        /* Initialize XLATEOBJ */
+        EXLATEOBJ_vInitialize(&exlo,
+                              ppalDIB,
+                              psurfDst->ppal,
+                              RGB(0xff, 0xff, 0xff),
+                              pdc->pdcattr->crBackgroundClr,
+                              pdc->pdcattr->crForegroundClr);
+
+        /* Perform the stretch operation */
+        IntEngStretchBlt(&psurfDst->SurfObj,
+                                   &psurfTmp->SurfObj,
+                                   NULL,
+                                   (CLIPOBJ *)&pdc->co,
+                                   &exlo.xlo,
+                                   &pdc->dclevel.ca,
+                                   &rcDst,
+                                   &rcSrc,
+                                   NULL,
+                                   &pdc->eboFill.BrushObject,
+                                   NULL,
+                                   WIN32_ROP3_TO_ENG_ROP4(dwRop));
+
+        /* Cleanup */
+        DC_vFinishBlit(pdc, NULL);
+        EXLATEOBJ_vCleanup(&exlo);
+    cleanup:
+        if (ppalDIB) PALETTE_ShareUnlockPalette(ppalDIB);
+        if (psurfTmp) SURFACE_ShareUnlockSurface(psurfTmp);
+        if (hbmTmp) GreDeleteObject(hbmTmp);
+        if (pdc) DC_UnlockDc(pdc);
+        if (safeBits) ExFreePoolWithTag(safeBits, TAG_DIB);
+
+    }
+
+    /* This is not what MSDN says is returned from this function, but it
```

Comment:
Do we have a test for this?

Reply:
Yes. It is tested in gdi32:dib and I have added this fact in the comment.